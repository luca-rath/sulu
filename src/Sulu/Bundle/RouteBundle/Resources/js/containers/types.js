// @flow

export type PageTreeRouteValue = {|
    page: ?string,
    suffix: ?string,
|};
