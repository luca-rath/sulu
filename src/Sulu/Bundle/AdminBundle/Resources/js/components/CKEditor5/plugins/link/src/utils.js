// @flow
import Node from '@ckeditor/ckeditor5-engine/src/view/node';
import AttributeElement from '@ckeditor/ckeditor5-engine/src/view/attributeelement';
import {URL_REGEX} from '../../../../Url/Url';

const linkElementSymbol = Symbol('linkElement');

/* eslint-disable no-control-regex */
const ATTRIBUTE_WHITESPACES = /[\u0000-\u0020\u00A0\u1680\u180E\u2000-\u2029\u205f\u3000]/g;
/* eslint-enable */

/*
 * Returns `true` if a given view node is the link element.
 */
export function isLinkElement(node: Node): boolean {
    return (
        node.is('attributeElement') &&
        !!node.getCustomProperty(linkElementSymbol)
    );
}

/*
 * Creates link AttributeElement with provided `href` attribute.
 */
export function createLinkElement(
    writer: Object,
    tag: string,
    attributes: Object,
    internalAttribute: string
): AttributeElement {
    // Priority 5 - https://github.com/ckeditor/ckeditor5-link/issues/121.
    const linkElement = writer.createAttributeElement(
        tag,
        {...attributes},
        {priority: 5}
    );
    writer.setCustomProperty(linkElementSymbol, true, linkElement);
    writer.setCustomProperty(
        'internalAttribute',
        internalAttribute,
        linkElement
    );

    return linkElement;
}

/*
 * Returns a safe URL based on a given value.
 *
 * An URL is considered safe if it is safe for the user (does not contain any malicious code).
 *
 * If URL is considered unsafe, a simple `"#"` is returned.
 */
export function ensureSafeUrl(url: string): string {
    return isSafeUrl(url) ? url : '#';
}

/*
 * Checks whether the given URL is safe for the user (does not contain any malicious code).
 */
export function isSafeUrl(url: string): boolean {
    const normalizedUrl = url.replace(ATTRIBUTE_WHITESPACES, '');

    return !!normalizedUrl.match(URL_REGEX);
}
