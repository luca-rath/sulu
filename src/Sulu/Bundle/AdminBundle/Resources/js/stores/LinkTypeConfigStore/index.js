// @flow
import LinkTypeConfigStore from './LinkTypeConfigStore';

export default LinkTypeConfigStore;
