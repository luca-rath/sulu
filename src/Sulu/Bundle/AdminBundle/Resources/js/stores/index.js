// @flow
import localizationStore from './localizationStore';
import type {Localization} from './localizationStore/types';
import MultiSelectionStore from './MultiSelectionStore';
import ResourceListStore from './ResourceListStore';
import ResourceStore from './ResourceStore';
import SingleSelectionStore from './SingleSelectionStore';
import userStore from './userStore';

export {
    MultiSelectionStore,
    ResourceListStore,
    ResourceStore,
    SingleSelectionStore,
    localizationStore,
    userStore,
};

export type {
    Localization,
};
