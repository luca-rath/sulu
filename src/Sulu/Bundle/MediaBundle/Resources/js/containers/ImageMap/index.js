// @flow
import ImageMap from './ImageMap';

export default ImageMap;
