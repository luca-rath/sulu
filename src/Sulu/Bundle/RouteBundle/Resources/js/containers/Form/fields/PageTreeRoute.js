// @flow
import React from 'react';
import type {FieldTypeProps} from 'sulu-admin-bundle/containers/Form/types';
import {observer} from 'mobx-react';
import {action, observable} from 'mobx';
import type {IObservableValue} from 'mobx';
import userStore from 'sulu-admin-bundle/stores/userStore';
import {Grid} from 'sulu-admin-bundle/components';
import SingleSelection from 'sulu-admin-bundle/containers/SingleSelection';
import ResourceLocator from 'sulu-admin-bundle/components/ResourceLocator';
import ResourceLocatorHistory from 'sulu-admin-bundle/containers/ResourceLocatorHistory';
import resourceLocatorStyles from 'sulu-admin-bundle/containers/Form/fields/resourceLocator.scss';
import {translate} from 'sulu-admin-bundle/utils';
import type {PageTreeRouteValue} from '../../types.js';

type Props = FieldTypeProps<?PageTreeRouteValue>;

@observer
class PageTreeRoute extends React.Component<Props> {
    @observable mode: string;

    constructor(props: Props): void {
        super(props);

        const {
            fieldTypeOptions: {
                modeResolver,
            },
        } = props;

        if (!modeResolver) {
            throw new Error('The "modeResolver" must be a function returning a promise with the desired mode');
        }

        modeResolver(props).then(action((mode) => this.mode = mode));
    }

    get locale(): IObservableValue<string> {
        const {formInspector} = this.props;

        return formInspector.locale ? formInspector.locale : observable.box(userStore.contentLocale);
    }

    get showHistory(): boolean {
        const {
            formInspector: {
                id,
            },
            fieldTypeOptions: {
                historyResourceKey,
                options: {
                    history,
                },
            },
        } = this.props;

        return !!history && !!id && !!historyResourceKey;
    }

    get pageValue(): ?string {
        const {
            value: {
                page: {
                    uuid,
                } = {
                    uuid: null,
                },
            } = {
                page: {
                    uuid: null,
                },
            },
        } = this.props;

        return uuid;
    }

    get suffixValue(): ?string {
        const {
            value: {
                suffix,
            } = {
                suffix: null,
            },
        } = this.props;

        return suffix;
    }

    handlePageChange = (value: ?string, page: ?Object = {
        path: null,
    }): void => {
        this.handleChange({
            ...this.props.value,
            page: {
                uuid: value,
                path: page.path,
            },
        });
    };

    handleSuffixChange = (value: ?string): void => {
        this.handleChange({
            ...this.props.value, suffix: value,
        });
    };

    handleChange = (value: ?PageTreeRouteValue) => {
        const {onChange} = this.props;

        onChange(value);
    };

    render() {
        if (!this.mode) {
            return null;
        }

        const {
            dataPath,
            disabled,
            formInspector: {
                id,
                resourceKey,
            },
            fieldTypeOptions: {
                historyResourceKey,
                options,
            },
        } = this.props;

        return (
            <>
                <Grid>
                    <Grid.Item colSpan={this.showHistory ? 5 : 6}>
                        <SingleSelection
                            adapter="column_list"
                            disabled={!!disabled}
                            displayProperties={['url']}
                            emptyText={translate('sulu_page.no_page_selected')}
                            icon="su-document"
                            listKey="pages"
                            locale={this.locale}
                            onChange={this.handlePageChange}
                            overlayTitle={translate('sulu_page.single_selection_overlay_title')}
                            resourceKey="pages"
                            value={this.pageValue}
                        />
                    </Grid.Item>

                    <Grid.Item colSpan={this.showHistory ? 7 : 6}>
                        <div className={resourceLocatorStyles.resourceLocatorContainer}>
                            <div className={resourceLocatorStyles.resourceLocator}>
                                <ResourceLocator
                                    disabled={!!disabled}
                                    id={dataPath}
                                    mode={this.mode}
                                    onChange={this.handleSuffixChange}
                                    value={this.suffixValue}
                                />
                            </div>

                            {this.showHistory &&
                                <div className={resourceLocatorStyles.resourceLocatorHistory}>
                                    <ResourceLocatorHistory
                                        id={id}
                                        options={{
                                            locale: this.locale,
                                            resourceKey: resourceKey,
                                            ...options,
                                        }}
                                        resourceKey={historyResourceKey}
                                    />
                                </div>
                            }
                        </div>
                    </Grid.Item>
                </Grid>
            </>
        );
    }
}

export default PageTreeRoute;
