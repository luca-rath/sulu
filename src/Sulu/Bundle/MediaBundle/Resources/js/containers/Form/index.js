// @flow
import MediaSelection from './fields/MediaSelection';
import MediaVersionUpload from './fields/MediaVersionUpload';
import SingleMediaUpload from './fields/SingleMediaUpload';
import SingleMediaSelection from './fields/SingleMediaSelection';
import ImageMap from './fields/ImageMap';

export {
    MediaSelection,
    MediaVersionUpload,
    SingleMediaUpload,
    SingleMediaSelection,
    ImageMap,
};
