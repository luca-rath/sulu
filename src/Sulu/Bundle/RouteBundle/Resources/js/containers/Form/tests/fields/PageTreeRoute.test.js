// @flow
import React from 'react';
import {mount} from 'enzyme';
import {FormInspector, ResourceFormStore, ResourceLocatorHistory, SingleSelection} from 'sulu-admin-bundle/containers';
import {fieldTypeDefaultProps} from 'sulu-admin-bundle/utils/TestHelper';
import {ResourceStore} from 'sulu-admin-bundle/stores';
import ResourceLocator from 'sulu-admin-bundle/components/ResourceLocator';
import {extendObservable as mockExtendObservable, observable} from 'mobx';
import SingleSelectionStore from 'sulu-admin-bundle/stores/SingleSelectionStore';
import PageTreeRoute from '../../fields/PageTreeRoute';

jest.mock('sulu-admin-bundle/utils/Translator', () => ({
    translate: jest.fn((key) => key),
}));

jest.mock('sulu-admin-bundle/stores/userStore', () => ({
    contentLocale: 'de',
}));

jest.mock('sulu-admin-bundle/containers/List/stores/ListStore', () => jest.fn(function() {
    this.selections = [];
    this.clearSelection = jest.fn();
    this.select = jest.fn();
}));

jest.mock(
    'sulu-admin-bundle/stores/ResourceStore',
    () => jest.fn(function(resourceKey, id, options) {
        this.resourceKey = resourceKey;
        this.id = id;

        if (options) {
            this.locale = options.locale;
        }
    })
);

jest.mock('sulu-admin-bundle/containers/Form/stores/MemoryFormStore', () => jest.fn(function(data, schema) {
    this.data = data;
    this.schema = schema;
    this.change = jest.fn().mockImplementation((name, value) => {
        this.data[name] = value;
    });
    this.validate = jest.fn().mockReturnValue(true);
    this.destroy = jest.fn();
}));

jest.mock(
    'sulu-admin-bundle/containers/Form/stores/ResourceFormStore',
    () => jest.fn(function(resourceStore, formKey) {
        this.resourceKey = resourceStore.resourceKey;
        this.id = resourceStore.id;
        this.locale = resourceStore.locale;

        if (formKey) {
            this.formKey = formKey;
        }
    })
);

jest.mock(
    'sulu-admin-bundle/containers/Form/FormInspector',
    () => jest.fn(function(resourceFormStore) {
        this.id = resourceFormStore.id;
        this.locale = resourceFormStore.locale;
        this.isFieldModified = jest.fn();
    })
);

jest.mock('sulu-admin-bundle/stores/SingleSelectionStore', () => jest.fn(function() {
    this.set = jest.fn((item) => {
        this.item = item;
    });
    this.loadItem = jest.fn((id) => {
        this.item = {id, url: '/test/' + id};
    });
    this.clear = jest.fn();

    mockExtendObservable(this, {
        item: undefined,
        loading: false,
    });
}));

const modePromiseValue = 'leaf';
const modePromise = Promise.resolve(modePromiseValue);
const modeResolver = jest.fn().mockImplementation(() => modePromise);

const fieldTypeOptions = {
    historyResourceKey: 'routes',
    modeResolver: modeResolver,
    options: {history: true},
};

const value = {
    page: {
        uuid: 'uuid-uuid-uuid-uuid',
    },
    suffix: '/hello',
};

test('Render a PageTreeRoute', () => {
    const locale = observable.box('de');
    const formInspector = new FormInspector(new ResourceFormStore(
        new ResourceStore('pages', 'diuu-diuu-diuu-diuu', {locale}),
        'test'
    ));

    const pageTreeRoute = mount(
        <PageTreeRoute
            {...fieldTypeDefaultProps}
            fieldTypeOptions={fieldTypeOptions}
            formInspector={formInspector}
            value={value}
        />
    );

    expect(modeResolver).toHaveBeenCalled();

    return modePromise.then(() => {
        pageTreeRoute.update();
        expect(pageTreeRoute.render()).toMatchSnapshot();
        expect(pageTreeRoute.find(SingleSelection).prop('value')).toBe(value.page.uuid);
        expect(SingleSelectionStore).toHaveBeenCalledWith('pages', 'uuid-uuid-uuid-uuid', locale, undefined);

        const singleSelection = pageTreeRoute.find(SingleSelection);

        singleSelection.instance().singleSelectionStore.item = {};
        singleSelection.update();

        expect(singleSelection.find('.item').text()).toBe('/test/uuid-uuid-uuid-uuid');
        expect(singleSelection.render()).toMatchSnapshot();

        expect(pageTreeRoute.find(ResourceLocator).prop('value')).toBe(value.suffix);
        expect(pageTreeRoute.find(ResourceLocator).prop('mode')).toBe(modePromiseValue);
        expect(pageTreeRoute.find(ResourceLocatorHistory).prop('id')).toBe('diuu-diuu-diuu-diuu');
    });
});

test('Render a PageTreeRoute without history', () => {
    const formInspector = new FormInspector(new ResourceFormStore(new ResourceStore('pages'), 'test'));

    const pageTreeRoute = mount(
        <PageTreeRoute
            {...fieldTypeDefaultProps}
            fieldTypeOptions={{
                ...fieldTypeOptions,
                options: {
                    history: false,
                },
            }}
            formInspector={formInspector}
            value={value}
        />
    );

    expect(modeResolver).toHaveBeenCalled();

    return modePromise.then(() => {
        pageTreeRoute.update();
        expect(pageTreeRoute.render()).toMatchSnapshot();
        expect(pageTreeRoute.find(ResourceLocatorHistory).length).toBe(0);
    });
});

test('Render a PageTreeRoute without value', () => {
    const formInspector = new FormInspector(new ResourceFormStore(new ResourceStore('pages'), 'test'));

    const pageTreeRoute = mount(
        <PageTreeRoute
            {...fieldTypeDefaultProps}
            fieldTypeOptions={fieldTypeOptions}
            formInspector={formInspector}
            value={null}
        />
    );

    expect(modeResolver).toHaveBeenCalled();

    return modePromise.then(() => {
        pageTreeRoute.update();
        expect(pageTreeRoute.render()).toMatchSnapshot();
        expect(pageTreeRoute.find(SingleSelection).prop('value')).toBe(null);
        expect(pageTreeRoute.find(ResourceLocator).prop('value')).toBe(null);
    });
});
