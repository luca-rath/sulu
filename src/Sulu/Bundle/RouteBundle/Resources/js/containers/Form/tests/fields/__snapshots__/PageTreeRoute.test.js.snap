// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a PageTreeRoute 1`] = `
<div
  class="grid"
>
  <div
    class="item colSpan colSpan-5 space-before-0 space-after-0"
  >
    <div
      class="singleItemSelection"
    >
      <button
        class="button left"
        type="button"
      >
        <span
          aria-label="su-document"
          class="su-document icon"
        />
      </button>
      <div
        class="itemContainer"
      >
        <div
          class="item"
        >
          <div
            class="empty"
          >
            sulu_page.no_page_selected
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="item colSpan colSpan-7 space-before-0 space-after-0"
  >
    <div
      class="resourceLocatorContainer"
    >
      <div
        class="resourceLocator"
      >
        <div
          class="resourceLocator"
        >
          <span
            class="fixed"
          >
            /
          </span>
          <label
            class="input default left"
          >
            <input
              id="/"
              type="text"
              value="hello"
            />
          </label>
        </div>
      </div>
      <div
        class="resourceLocatorHistory"
      >
        <button
          class="button link"
          type="button"
        >
          <span
            aria-label="su-process"
            class="su-process buttonIcon"
          />
          <span
            class="text"
          >
            sulu_admin.show_history
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Render a PageTreeRoute 2`] = `
<div
  class="singleItemSelection"
>
  <button
    class="button left"
    type="button"
  >
    <span
      aria-label="su-document"
      class="su-document icon"
    />
  </button>
  <div
    class="itemContainer"
  >
    <div
      class="item"
    >
      <div>
        <span
          class="itemColumn"
          style="width: 100%;"
        >
          /test/uuid-uuid-uuid-uuid
        </span>
      </div>
    </div>
    <button
      class="removeButton"
      type="button"
    >
      <span
        aria-label="su-trash-alt"
        class="su-trash-alt"
      />
    </button>
  </div>
</div>
`;

exports[`Render a PageTreeRoute without history 1`] = `
<div
  class="grid"
>
  <div
    class="item colSpan colSpan-6 space-before-0 space-after-0"
  >
    <div
      class="singleItemSelection"
    >
      <button
        class="button left"
        type="button"
      >
        <span
          aria-label="su-document"
          class="su-document icon"
        />
      </button>
      <div
        class="itemContainer"
      >
        <div
          class="item"
        >
          <div
            class="empty"
          >
            sulu_page.no_page_selected
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="item colSpan colSpan-6 space-before-0 space-after-0"
  >
    <div
      class="resourceLocatorContainer"
    >
      <div
        class="resourceLocator"
      >
        <div
          class="resourceLocator"
        >
          <span
            class="fixed"
          >
            /
          </span>
          <label
            class="input default left"
          >
            <input
              id="/"
              type="text"
              value="hello"
            />
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Render a PageTreeRoute without value 1`] = `
<div
  class="grid"
>
  <div
    class="item colSpan colSpan-6 space-before-0 space-after-0"
  >
    <div
      class="singleItemSelection"
    >
      <button
        class="button left"
        type="button"
      >
        <span
          aria-label="su-document"
          class="su-document icon"
        />
      </button>
      <div
        class="itemContainer"
      >
        <div
          class="item"
        >
          <div
            class="empty"
          >
            sulu_page.no_page_selected
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="item colSpan colSpan-6 space-before-0 space-after-0"
  >
    <div
      class="resourceLocatorContainer"
    >
      <div
        class="resourceLocator"
      >
        <div
          class="resourceLocator"
        >
          <span
            class="fixed"
          >
            /
          </span>
          <label
            class="input default left"
          >
            <input
              id="/"
              type="text"
              value=""
            />
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;
